### PR TITLE
Channel participation: implement remove channel for application channels

### PR DESCRIPTION
**Pull request by wlahti**
#### Type of change

- New feature
- Test update

#### Description

This stacks on top of https://github.com/hyperledger/fabric/pull/1628. It implements the registrar's RemoveChannel function and includes a basic integration test.

#### Related issues

[FAB-17964](https://jira.hyperledger.org/browse/FAB-17964) - implement registrar RemoveChannel
[FAB-18093](https://jira.hyperledger.org/browse/FAB-18093) - integration test

### CODE REVIEW

File: orderer/common/multichannel/registrar.go [hunk 1]
```diff
@@ -581,11 +581,42 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
...
+
 	if r.SystemChannelID() != "" && channelID != r.SystemChannelID() {
 		return types.ErrSystemChannelExists
 	}
-	return errors.New("Not implemented yet")
+
+	cs, ok := r.chains[channelID]
+	if !ok {
+		return types.ErrChannelNotExist
+	}
```
Comment (tock-ibm) [inline orderer/common/multichannel/registrar.go:615]: a follower may also be removed, not just a chain.

File: orderer/common/multichannel/registrar.go [hunk 2]
```diff
@@ -581,11 +581,42 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
...
+	if !ok {
+		return errors.Errorf("cannot remove non-raft application channel %s", channelID)
+	}
+
+	raftChain.HaltExternal()
+
+	err := r.ledgerFactory.Remove(channelID)
+	if err != nil {
+		return errors.Errorf("error removing ledger for channel %s", channelID)
+	}
```
Comment (tock-ibm) [inline orderer/common/multichannel/registrar.go:631]: This may take a long time for a big chain, and blocks the request thread. We need to benchmark the Remove  and see if it is a real issue. If it is, put a Jira out to implement this asynch to the calling thread. That is, spin a go-routine to do it, and in the meantime put the channel in a "quarantine" map. In quarantine, a channel cannot be joined or deleted again until it finishes deleting its storage.

File: orderer/common/multichannel/registrar.go [hunk 3]
```diff
@@ -581,11 +581,42 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
...
+
+	// check if chain is etcdraft by looking for the HaltExternal function, which
+	// is only implemented by raft chains
+	raftChain, ok := cs.Chain.(externalHalter)
+	if !ok {
+		return errors.Errorf("cannot remove non-raft application channel %s", channelID)
+	}
+
+	raftChain.HaltExternal()
```
Comment (tock-ibm) [inline orderer/common/multichannel/registrar.go]: Not sure we need `HaltExternal`, I think a simple fix in the raft chain can let us use `Halt`.

File: orderer/common/multichannel/registrar.go (hunk 1 above)
Comment (wlahti) [reply orderer/common/multichannel/registrar.go:615]: Done. Haven't added the tests yet (just noticed your join channel as follower PR was merged so I'll use that as a reference).

File: orderer/common/multichannel/registrar_test.go [hunk 4]
```diff
@@ -806,6 +805,96 @@ func TestRegistrar_RemoveChannel(t *testing.T) {
...
+			info, err := registrar.JoinChannel("my-raft-channel", genesisBlockAppRaft, true)
+			require.NoError(t, err)
+			require.Equal(t, types.ChannelInfo{Name: "my-raft-channel", URL: "", ClusterRelation: "member", Status: "active", Height: 0x1}, info)
+			// After creating the chain, it exists
+			require.NotNil(t, registrar.GetChain("my-raft-channel"))
+
+			err = registrar.RemoveChannel("my-raft-channel", false)
+			require.NoError(t, err)
+			// After deleting the chain, it no longer exists
+			require.Nil(t, registrar.GetChain("my-raft-channel"))
```
Comment (tock-ibm) [inline orderer/common/multichannel/registrar_test.go]: Check that ledger resources were removed as well.

File: orderer/common/multichannel/util_test.go [hunk 5]
```diff
@@ -31,6 +31,10 @@ func (c *mockChainCluster) StatusReport() (types.ClusterRelation, types.Status)
 	return types.ClusterRelationMember, types.StatusActive
 }
 
+func (c *mockChainCluster) HaltExternal() {
+	c.mockChain.Halt()
+}
+
```
Comment (tock-ibm) [inline orderer/common/multichannel/util_test.go]: Why is this needed?

File: orderer/common/multichannel/registrar.go [hunk 6]
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
...
 func (r *Registrar) RemoveChannel(channelID string, removeStorage bool) error {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
 	if r.SystemChannelID() != "" && channelID != r.SystemChannelID() {
 		return types.ErrSystemChannelExists
 	}
-	return errors.New("Not implemented yet")
+
+	cs, ok := r.chains[channelID]
```
Comment (tock-ibm) [inline orderer/common/multichannel/registrar.go:612]: Add a check right above here that rejects the removal of the system channel, and returns "not yet implemented". Removal of the system channel is a different (Jira) story that requires a special process, since it mandates the conversion of every chain life-cycle from etcdraft+inactive to etcdraft+follower. This may require a restart of the orderer or the restart of every chain, as we'll decide when we tackle that story.

File: orderer/common/multichannel/registrar.go [hunk 7]
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
...
+}
+
+// validateRaftChain returns an error if the chain is not a raft chain.
+func validateRaftChain(channelID string, chain consensus.StatusReporter) error {
+	clusterRelation, _ := chain.StatusReport()
+	if clusterRelation == types.ClusterRelationNone {
+		return errors.Errorf("cannot remove non-raft application channel %s", channelID)
+	}
+
+	return nil
```
Comment (tock-ibm) [inline orderer/common/multichannel/registrar.go]: I am thinking, there is nothing in the "Remove" that is specific to raft/follower... We call `Halt`, we remove the ledger, and that is it (unlike the "Join"...). So perhaps we can make the code simpler by removing this check?

File: orderer/common/multichannel/util_test.go (hunk 5 above)
Comment (wlahti) [reply orderer/common/multichannel/util_test.go]: Good catch. This is leftover from the previous implementation in the halt PR.  I'll remove this.

File: orderer/common/multichannel/registrar_test.go (hunk 4 above)
Comment (wlahti) [reply orderer/common/multichannel/registrar_test.go]: Done

File: orderer/common/multichannel/registrar.go (hunk 7 above)
Comment (wlahti) [reply orderer/common/multichannel/registrar.go]: Done

File: orderer/common/multichannel/registrar.go [hunk 8]
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
 // RemoveChannel instructs the orderer to remove a channel.
 // Depending on the removeStorage parameter, the storage resources are either removed or archived.
 func (r *Registrar) RemoveChannel(channelID string, removeStorage bool) error {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
```
[the remark this answers is not in the document]
Comment (wlahti) [reply orderer/common/multichannel/registrar.go]: Done

File: orderer/common/multichannel/registrar.go (hunk 6 above)
Comment (wlahti) [reply orderer/common/multichannel/registrar.go:612]: Done

File: orderer/common/multichannel/registrar.go [hunk 9]
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
...
+	if err != nil {
+		return err
+	}
+
+	cs.Halt()
+
+	err = r.ledgerFactory.Remove(channelID)
+	if err != nil {
+		return errors.Errorf("error removing ledger for channel %s", channelID)
+	}
```
[the remark this answers is not in the document]
Comment (tock-ibm) [reply orderer/common/multichannel/registrar.go]: Opened https://jira.hyperledger.org/browse/FAB-18132 for this.

File: orderer/common/multichannel/registrar.go (hunk 1 above)
Comment (wlahti) [reply orderer/common/multichannel/registrar.go:615]: Added the UT for this (a while ago).

File: integration/raft/channel_participation_test.go [hunk 10]
```diff
@@ -124,43 +125,89 @@ var _ = Describe("ChannelParticipation", func() {
...
-			channelparticipation.ListOne(network, orderer1, expectedChannelInfo)
+			channelparticipation.Join(network, orderer1, "another-participation-trophy", genesisBlock2, expectedChannelInfoAPT)
+			channelparticipation.ListOne(network, orderer1, expectedChannelInfoAPT)
 
-			By("listing all channels")
+			By("listing all channels for orderer1")
 			channelparticipation.List(network, orderer1, []string{"participation-trophy", "another-participation-trophy"})
+
+			By("removing orderer1 from the consenter set")
+			channelConfig := nwo.GetConfig(network, peer, orderer2, "participation-trophy")
```
Comment (tock-ibm) [inline integration/raft/channel_participation_test.go:166]: This is indeed the recommended flow. First remove from the consenters set, then remove from orderer. However - I have two comments:
1. The reverse flow should be working as well, you just have to wait for the remaining consenters to re-elect a leader in case the leader was the one removed.
2. When a raft chain detects eviction, it should turn into a follower. This is not yet implemented, the chain will just print something and exit. After https://jira.hyperledger.org/browse/FAB-17915 that will change.  So when that happens one should also check that the orderer behaves as expected when evicted from a channel.

I would insert a TODO and/or a Jira in the integration test story to catch these two test cases and cover then later.

**Comment by tock-ibm** [comment]
Other than the two minor comments about future work above, this looks good.

File: integration/raft/channel_participation_test.go (hunk 10 above)
Comment (wlahti) [reply integration/raft/channel_participation_test.go:166]: Sounds good. I'lll open sub-tasks in the integration test task.

File: orderer/common/multichannel/registrar.go [hunk 11]
```diff
@@ -577,8 +577,56 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
...
+	delete(r.chains, channelID)
+
+	logger.Infof("Removed channel: %s", channelID)
+
+	return nil
+}
+
+func (r *Registrar) removeFollower(channelID string, follower *follower.Chain) error {
+	follower.Halt()
+
```
[the remark this answers is not in the document]
Comment (wlahti) [reply orderer/common/multichannel/registrar.go:642]: Added the TODO.

**Review by C0rWin — APPROVED** [review]
